text/0061-functions.md

```diff
@@ -0,0 +1,412 @@
+```
+def eq(?a, ?b) ?a == ?b ;
+
+permit(principal, action, resource)
+when {
+  eq(principal,resource.owner) ||
+  eq(principal.org,"Admin")
+};
+```
+
+Likewise, any static analysis tools would work via inlining.
+
+### Templates
+Macros use the same syntax for their parameter variables as do templates, since in both cases variables are "holes" that are filled in with a Cedar expression to make a full construct -- for templates, the construct is a policy, for macros it is an expression.
+
+Templates can include calls to macros. For example:
+
+```
+def semver(?major, ?minor, ?patch) ... // same as initial example
+def semverGT(?lhs, ?rhs) ... // same as initial example
+permit (
+  principal == ?principal,
+  action == Action::"view",
+  resource in ?resource)
+when {
+  semverGT(resource.apiVersion, semver(2,1,0))
+};
+```
+
+When we link this template, we get a full policy, as usual.
+
+A macro parameter can end up having the same name as a template variable, with the effect that it shadows it, which is expected with normal lexical scoping:
+
+```
+def isOwner(?principal,?resource) ?principal == ?resource.owner;
+
+permit(principal,action,resource in ?resource)
+when {
+  isOwner(principal,resource)
+};
+```
+
+Here, notice that the `isOwner` macro's reference to `?resource` is to its parameter, not to the template slot. (We'd recommend to users to choose different parameter names to avoid confusion.)
+
+As already mentioned, you cannot refer to a parameter in a macro that it does not bind, which enforces hygiene. So the following is not allowed.
+
+```
+def isOwner() ?principal == ?resource.owner; 
+// ERROR: cannot refer to unbound variables ?principal, ?resource
+```
+
+
+## Drawbacks
+
+### Redundancy
+Cedar macros can only accomplish things that can already accomplished with Cedar expressions.
+This means we are not expanding the expressive power of Cedar in any way. 
+We are also adding more than one way to accomplish the same task.
+Bringing back our SemVer example:
+```
+// Permit access when the api is greater than 2.1
+permit (principal, action == Action::"someAction", resource)
+when {
+    resource.apiVersion.major > 2 ||
+    (resource.apiVersion.major == 2 && resource.apiVersion.minor >= 1)
+};
+```
+This policy does accomplish the user's goal of encoding the SemVer relationship.
+The problem is readability. A reader of this policy has to work out that it is implementing the standard semantic version comparison operator, and not some bespoke versioning scheme. Another problem is maintainability. If you have multiple policies that reason about SemVers, you have to repeat the logic, inline, in each policy. This violates basic software engineering tenets (Don’t Repeat Yourself), allowing bugs to sneak in via typos or botched copy/paste.
```

Comment:
While this example is a bit more verbose it is serializable / deserializable across different cedar applications. How do we store these policies with macros? I see the management becoming a bit more complicated.

Reply:
The management burden shouldn't be underestimated.

As far as I see it, there are two options:

(1)  The service / application stores a representation of the policy with the macro calls.  Then, the caller is responsible for gathering all the necessary macro definitions and passing them to the authorizer, just as they are currently responsible for gathering all the relevant policies and entities and passing them to the authorizer.  This same burden extends to services that may need to ship policies and macros to a local agent.  

(2) The services / application stores a representation of the expanded policy---that is, all the macros are inlined before storing the policy.  This is how macros are handled in mature languages with macro systems. And it means that macros no longer need to be shipped.  However, it also becomes impossible to restore the original policy text from the expanded representation, and if there is a bug in the macro that is propagated to all policies that use the macro, it cannot be fixed by just fixing the macro:  the macro must be fixed and all policies that use it must be re-expanded.

Reply:
I think this slightly overstates the management burden.  The service / application is already storing policies, so also storing macros is not that huge of a lift.  @aaronjeline also talked to engineers from at least one managed Cedar service in writing this RFC, and they did not have concerns about the management burden.

Reply:
Thank you for the responses. In both cases it seems challenging as it assumes a policy is already written.

A common use case is the policy author != the policy evaluator. For example, a system administrator writes a policy in the browser via some typescript plugin. This policy is sent to some backend system in java that wants to validate it against a schema and store it in a persistent store. The actual pulling of the policy and evaluation is a different system (a rust system).

It isn't clear how I would support this cleanly with macros. Not this RFC needs to address those concerns but this is how I see Cedar being used in practice.